src/Modules/CrestApps.OrchardCore.AI.Chat/Hubs/AIChatHub.cs

```diff
@@ -550,4 +554,137 @@ private async Task ProcessUtilityAsync(ChannelWriter<CompletionPartialMessage> w
+            {
+                await Clients.Caller.ReceiveError(S["{0} is required.", nameof(prompt)].Value);
+                return;
+            }
+
+            var httpContext = Context.GetHttpContext();
+
+            if (!await _authorizationService.AuthorizeAsync(httpContext.User, AICustomChatPermissions.ManageOwnCustomChatInstances))
+            {
+                await Clients.Caller.ReceiveError(S["You are not authorized to use custom chat instances."].Value);
+                return;
+            }
+
+            var instance = await _customChatInstanceManager.FindByIdForCurrentUserAsync(instanceId);
+
+            if (instance == null)
+            {
+                await Clients.Caller.ReceiveError(S["Custom chat instance not found."].Value);
+                return;
+            }
+
+            await ProcessCustomChatPromptAsync(writer, instance, prompt.Trim(), history, cancellationToken);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "An error occurred while processing the custom chat prompt.");
+
+            var errorMessage = new CompletionPartialMessage
+            {
+                MessageId = IdGenerator.GenerateId(),
+                Content = GetFriendlyErrorMessage(ex).Value,
+            };
+
+            await writer.WriteAsync(errorMessage, cancellationToken);
+        }
+        finally
+        {
+            writer.Complete();
+        }
+    }
+
+    private async Task ProcessCustomChatPromptAsync(ChannelWriter<CompletionPartialMessage> writer, AICustomChatInstance instance, string prompt, IEnumerable<ChatMessageEntry> history, CancellationToken cancellationToken)
+    {
+        var messageId = IdGenerator.GenerateId();
+
+        var completionContext = new AICompletionContext
+        {
+            ConnectionName = instance.ConnectionName,
+            DeploymentId = instance.DeploymentId,
+            SystemMessage = instance.SystemMessage,
+            Temperature = instance.Temperature,
+            TopP = instance.TopP,
+            FrequencyPenalty = instance.FrequencyPenalty,
+            PresencePenalty = instance.PresencePenalty,
+            MaxTokens = instance.MaxTokens,
+            PastMessagesCount = instance.PastMessagesCount,
+            ToolNames = instance.ToolNames?.ToArray(),
+            UserMarkdownInResponse = true,
+        };
+
+        var references = new Dictionary<string, AICompletionReference>();
+
+        var builder = new StringBuilder();
+
+        // Build the transcript from history and current prompt
+        var transcript = new List<ChatMessage>();
+
+        if (history != null)
+        {
+            var pastMessagesCount = instance.PastMessagesCount ?? 10;
+            var recentHistory = history.TakeLast(pastMessagesCount);
+
+            foreach (var entry in recentHistory)
+            {
+                var role = entry.Role?.ToLowerInvariant() == "user" ? ChatRole.User : ChatRole.Assistant;
```

Comment:
I don't think we lower the role for comparison. I believe we already store it as ChatRole.